### PR TITLE
[charts] Allow configuring bar size

### LINKED ISSUE #9609: [charts] Improve band axis
Add options to customize offset and padding. For now, bar have a fix padding in percent of the bandWidth

## Echarts system

The bar series have some [properties](https://echarts.apache.org/en/option.html#series-bar.barWidth):

- `barWidth`/`barMinWidth`/`barMaxWidth`: explicit
- `barGap`: gap between two bars. If in percent, it's the percent of the bar width not the percent of the bandwidth
- `barCategoryGap`: the gap between the last bar of a category and the first bar of the next category

Those values are shared in a weird way: they are defined in series but shared among series. If two series define the same property the last definition wins

## [d3-scale (band)](https://github.com/d3/d3-scale#band-scales)

- `paddingInner`: ratio (between 0 and 1) of the band width and the space between two bands
- `paddingOuter`: defines the space between the extremities and the first/last band
- `padding`: the two above using a single function

## Idea

It seems to make more sense to put those definitions into the axis settings

We have `barCategoryGap` property which is equivalent to the `paddingInner` property
We could introduce 
- `bandGap` that would have same behavior as `paddingInner`
- `gap` that would have same behavior as `barGap`
- let the bar width automatically computed as `bandWidth * (1 - gap) / (nb of bars)` It does not allow bar overflow or various bar width but it does not seems to be an important point

### PR DESCRIPTION
**Pull request by alexfauquette**
Fix #9609

The DX can bee seen [here](https://deploy-preview-9632--material-ui-x.netlify.app/x/react-charts/bars/#bar-size)

### CODE REVIEW

File: packages/x-charts/src/BarChart/BarPlot.tsx [hunk 1]
```diff
@@ -1,9 +1,40 @@
...
 import { CartesianContext } from '../context/CartesianContextProvider';
-import { isBandScale } from '../internals/isBandScale';
 import { BarElement } from './BarElement';
+import { isBandScaleConfig } from '../models/axis';
 
+/**
+ * Solution of the equations
+ * W = barWidth * N + offset * (N-1)
+ * offset / (offset + barWidth) = r
+ * @param bandWidth The width available to put bars.
```
Comment (LukasTy) [inline packages/x-charts/src/BarChart/BarPlot.tsx:11]:
```suggestion
 * @param bandWidth The width available to place bars.
```

File: packages/x-charts/src/BarChart/BarPlot.tsx [hunk 2]
```diff
@@ -1,9 +1,40 @@
...
+ * @param numberOfGroups The number of bars to put in that space.
```
Comment (LukasTy) [inline packages/x-charts/src/BarChart/BarPlot.tsx:12]:
```suggestion
 * @param numberOfGroups The number of bars to place in that space.
```

File: docs/data/charts/bars/bars.md [hunk 3]
```diff
@@ -15,6 +15,20 @@ This axis might have `scaleType='band'` and its `data` should have the same leng
 
 {{"demo": "BasicBars.js"}}
 
+## Bar size
+
+You can define bar dimensions with `categoryGapRatio` and `barGapRatio` properties.
+
+The `categoryGapRatio` defines the gap between two categories.
+The ratio is obtained size of the gap divided by the size of the category (the space used by bars).
```
Comment (LukasTy) [inline docs/data/charts/bars/bars.md:23]:
```suggestion
The ratio is obtained by dividing the size of the gap by the size of the category (the space used by bars).
```

File: docs/data/charts/bars/bars.md [hunk 4]
```diff
@@ -15,6 +15,20 @@ This axis might have `scaleType='band'` and its `data` should have the same leng
...
+
+The `barGapRatio` defines the gap between two bars of same category.
```
Comment (LukasTy) [inline docs/data/charts/bars/bars.md:25]:
```suggestion
The `barGapRatio` defines the gap between two bars of the same category.
```

**Review by LukasTy — COMMENTED** [review]
Nice, great feature! ❤️ 💯 
Everything looks great, the demo is superb. 👍 
There is one functional nitpick: shouldn't the gap actually behave like a gap?
Currently the gap is only visual, the hovering effect is still applied only to the actual bar area, but the tooltip does activate even on the empty space (gap).
What do you think? 🤔 

https://github.com/mui/mui-x/assets/4941090/e487305f-ba87-4b5c-96fc-2050905da8cb

**Comment by LukasTy** [comment]
Yes, that would also work and if it's easier to do - let's go for it. 👍

**Review by LukasTy — APPROVED** [review]
👏 🚀

**Comment by ThreeCrown** [comment]
How does someone manipulate the outside padding of the chart though? Great for the bars though?

image

**Comment by alexfauquette** [comment]
@ThreeCrown I don't understand what you're trying to do. Could you open an issue with an example of what you tried and what you expect?

**Pull request merged**